modules/byDataAnalyticsAdapter.js

```diff
@@ -0,0 +1,309 @@
+    switch (eventType) {
+      case NO_BID:
+        onNoBidData(args);
+        break;
+      case BID_TIMEOUT:
+        onBidTimeout(args);
+        break;
+      case AUCTION_END:
+        onAuctionEnd(args);
+        break;
+      default:
+        break;
+    }
+  }
+});
+
+// save the base class function
+ascAdapter.originEnableAnalytics = ascAdapter.enableAnalytics;
+// override enableAnalytics so we can get access to the config passed in from the page
+ascAdapter.enableAnalytics = function(config) {
+  if (this.initConfig(config)) {
+    logInfo('initiated:', initOptions);
+    initOptions.isCorrectOption && ascAdapter.getVisitorData();
+    ascAdapter.originEnableAnalytics(config);
+  }
+};
+
+ascAdapter.initConfig = function (config) {
+  let isCorrectOption = true;
+  initOptions = {};
+  logInfo('initConfig', config);
+  initOptions.options = utils.deepClone(config.options);
+  initOptions.clientId = initOptions.options.clientId || null;
+  initOptions.logFrequency = initOptions.options.logFrequency;
+  if (!initOptions.clientId) {
+    logError('"options.clientId" should not empty!!');
+    isCorrectOption = false;
+  }
+  initOptions.isCorrectOption = isCorrectOption;
+  this.initOptions = initOptions;
+  return isCorrectOption;
+};
+
+ascAdapter.getVisitorData = function(data = {}) {
+  var ua = data.userId ? data : {};
+  var module = {
+    options: [],
+    header: [window.navigator.platform, window.navigator.userAgent, window.navigator.appVersion, window.navigator.vendor, window.opera],
+    dataos: [
+      { name: 'Windows Phone', value: 'Windows Phone', version: 'OS' },
+      { name: 'Windows', value: 'Win', version: 'NT' },
+      { name: 'iPhone', value: 'iPhone', version: 'OS' },
+      { name: 'iPad', value: 'iPad', version: 'OS' },
+      { name: 'Kindle', value: 'Silk', version: 'Silk' },
+      { name: 'Android', value: 'Android', version: 'Android' },
+      { name: 'PlayBook', value: 'PlayBook', version: 'OS' },
+      { name: 'BlackBerry', value: 'BlackBerry', version: '/' },
+      { name: 'Macintosh', value: 'Mac', version: 'OS X' },
+      { name: 'Linux', value: 'Linux', version: 'rv' },
+      { name: 'Palm', value: 'Palm', version: 'PalmOS' }
+    ],
+    databrowser: [
+      { name: 'Chrome', value: 'Chrome', version: 'Chrome' },
+      { name: 'Firefox', value: 'Firefox', version: 'Firefox' },
+      { name: 'Safari', value: 'Safari', version: 'Version' },
+      { name: 'Internet Explorer', value: 'MSIE', version: 'MSIE' },
+      { name: 'Opera', value: 'Opera', version: 'Opera' },
+      { name: 'BlackBerry', value: 'CLDC', version: 'CLDC' },
+      { name: 'Mozilla', value: 'Mozilla', version: 'Mozilla' }
+    ],
+    init: function () {
+      var agent = this.header.join(' '); var os = this.matchItem(agent, this.dataos); var browser = this.matchItem(agent, this.databrowser);
+      return { os: os, browser: browser };
+    },
+    matchItem: function (string, data) {
```

Comment:
@Prebid-bydata , wouldn't it be better to do the whole user-agent parsing on the server?

Reply:
hi @aleksatr thinking on this,  it tales micro seconds to execute.